rclcpp/include/rclcpp/strategies/allocator_memory_strategy.hpp

```diff
@@ -187,7 +187,7 @@ class AllocatorMemoryStrategy : public memory_strategy::MemoryStrategy
         group->find_waitable_ptrs_if(
           [this](const rclcpp::Waitable::SharedPtr & waitable) {
             waitable_handles_.push_back(waitable);
-            return false;
+            return true;
```

Comment:
I don't think this is extending the lifetime of the waitable, it is just stopping the iteration through the waitables after visiting the first one:

https://github.com/ros2/rclcpp/blob/2aaeee72a6efa3475837245869eb4128f41a57fd/rclcpp/include/rclcpp/callback_group.hpp#L155

So I think you're just avoiding the bug, not fixing it with this.

These functions are being used for their side effects, essentially always returning false means the "find" function will be called for each entry, it's essentially being used as a "for each" or "map" like function. It's kind of ugly to be honest... Not sure if it's my fault or not, haha. Doesn't matter though.

Reply:
> So I think you're just avoiding the bug, not fixing it with this.

Yeah, you are completely right about this.  I misunderstood the logic.

Reply:
[Waitable](https://github.com/ros2/rclcpp/blob/01a6befdde7f78b6f7788e6180cc5ed38390e955/rclcpp/include/rclcpp/waitable.hpp#L26) class doesn't define a virtual destructor, and then:

https://github.com/ros2/rclcpp/blob/01a6befdde7f78b6f7788e6180cc5ed38390e955/rclcpp/include/rclcpp/qos_event.hpp#L79-L83

Sounds like that might be the problem (I really don't know if that's the case, just wondering)

Reply:
> Sounds like that might be the problem (I really don't know if that's the case, just wondering)

Huh, good point.  It does look like this is *a* problem, but putting a virtual destructor on the `Waitable` class just moved the problem around.  More debugging required.

Reply:
From the discussion in yesterday's meeting, I have an idea I want to share that might help to fix the problem:

Events can be created both from a [publisher](https://github.com/ros2/rmw/blob/d030ad19eaeb6bee3eda7c0197360442e2db0a7a/rmw/include/rmw/event.h#L79) and a [subscription](https://github.com/ros2/rmw/blob/d030ad19eaeb6bee3eda7c0197360442e2db0a7a/rmw/include/rmw/event.h#L97).
The `rmw` implementation then holds a pointer to underlying implementation the Publisher/Subscription (see [1](https://github.com/ros2/rmw_connext/blob/c90e7752a2f482aef4afe53dc0ce7f45c632152e/rmw_connext_cpp/src/rmw_event.cpp#L33), [2](https://github.com/ros2/rmw_connext/blob/c90e7752a2f482aef4afe53dc0ce7f45c632152e/rmw_connext_cpp/src/rmw_event.cpp#L47), [3](https://github.com/ros2/rmw_connext/blob/c90e7752a2f482aef4afe53dc0ce7f45c632152e/rmw_connext_shared_cpp/src/event.cpp#L46)).

Then, at `rclcpp` level, the created `EventHandler` is not holding the `Publisher`/`Subscription`:

https://github.com/ros2/rclcpp/blob/2aaeee72a6efa3475837245869eb4128f41a57fd/rclcpp/include/rclcpp/qos_event.hpp#L101-L102

when that actually should be hold.

It's not possible to hold the `rclcpp` objects in the event handlers, as that would create a cyclic dependency ([4](https://github.com/ros2/rclcpp/blob/2aaeee72a6efa3475837245869eb4128f41a57fd/rclcpp/include/rclcpp/subscription_base.hpp#L286), [5](https://github.com/ros2/rclcpp/blob/2aaeee72a6efa3475837245869eb4128f41a57fd/rclcpp/include/rclcpp/publisher_base.hpp#L218)), but holding the `rcl` objects should solve the problem.

To avoid creating derived classes of [QOSEventHandler](https://github.com/ros2/rclcpp/blob/2aaeee72a6efa3475837245869eb4128f41a57fd/rclcpp/include/rclcpp/qos_event.hpp#L106) for publisher and subscriptions, we can store a `shared_ptr<void>` to the parent handler.

Disclaimer: I haven't tested the issue and then I'm not sure if this is actually the root cause of the problem, but there is definitely something "unsound" here.

Reply:
Thanks for the thoughts @ivanpauno .  You've got the problem exactly right; we aren't holding a reference to the Publisher, and so some of the underlying rcl structures can get freed and then improperly reused in the executor.

I like your idea about holding onto a shared pointer for the underlying rcl structures as a fix.  I'll start trying to do that and see what happens there.

Reply:
All right, this fix works, and I feel a *lot* better about it in general.  @ivanpauno thanks again for the idea.  @wjwwood another review here would be appreciated.